make/Images.gmk

```diff
@@ -85,29 +86,97 @@ JLINK_TOOL := $(JLINK) -J-Djlink.debug=true \
 
+DEFAULT_JDK_IMAGE_SUPPORT_DIR := $(SUPPORT_OUTPUTDIR)/images/jdk
+ifeq ($(JLINK_PRODUCE_RUNTIME_LINK_JDK), true)
+  JDK_LINK_OUTPUT_DIR := $(SUPPORT_OUTPUTDIR)/images/runtime-link-initial-jdk
+  JDK_IMAGE_SUPPORT_DIR := $(SUPPORT_OUTPUTDIR)/images/runtime-link-support
+else
+  JDK_LINK_OUTPUT_DIR := $(JDK_IMAGE_DIR)
+  JDK_IMAGE_SUPPORT_DIR := $(DEFAULT_JDK_IMAGE_SUPPORT_DIR)
+endif
+
 ifeq ($(JLINK_KEEP_PACKAGED_MODULES), true)
-  JLINK_JDK_EXTRA_OPTS := --keep-packaged-modules $(JDK_IMAGE_DIR)/jmods
+  ifeq ($(JLINK_PRODUCE_RUNTIME_LINK_JDK), true)
+    JLINK_RUNTIME_CREATE_EXTRA := --keep-packaged-modules $(JDK_IMAGE_DIR)/jmods
+  endif
+  JLINK_JDK_EXTRA_OPTS := --keep-packaged-modules $(JDK_LINK_OUTPUT_DIR)/jmods
 endif
 
 JLINK_DISABLE_WARNINGS := | ( $(GREP) -v -e "WARNING: Using incubator module" || test "$$?" = "1" )
 
-JDK_IMAGE_SUPPORT_DIR := $(SUPPORT_OUTPUTDIR)/images/jdk
 JRE_IMAGE_SUPPORT_DIR := $(SUPPORT_OUTPUTDIR)/images/jre
 
 $(eval $(call SetupExecute, jlink_jdk, \
     WARN := Creating jdk image, \
     DEPS := $(JDK_JMODS) $(BASE_RELEASE_FILE) \
         $(call DependOnVariable, JDK_MODULES_LIST, $(JDK_IMAGE_SUPPORT_DIR)/_jlink_jdk.vardeps), \
-    OUTPUT_DIR := $(JDK_IMAGE_DIR), \
+    OUTPUT_DIR := $(JDK_LINK_OUTPUT_DIR), \
     SUPPORT_DIR := $(JDK_IMAGE_SUPPORT_DIR), \
-    PRE_COMMAND := $(RM) -r $(JDK_IMAGE_DIR), \
+    PRE_COMMAND := $(RM) -r $(JDK_LINK_OUTPUT_DIR), \
     COMMAND := $(JLINK_TOOL) --add-modules $(JDK_MODULES_LIST) \
-        $(JLINK_JDK_EXTRA_OPTS) --output $(JDK_IMAGE_DIR) \
+        $(JLINK_JDK_EXTRA_OPTS) --output $(JDK_LINK_OUTPUT_DIR) \
         $(JLINK_DISABLE_WARNINGS), \
 ))
 
 JLINK_JDK_TARGETS := $(jlink_jdk)
 
+ifeq ($(JLINK_PRODUCE_RUNTIME_LINK_JDK), true)
+
+  RL_BUILD_MODULE_NAME := jdk.unsupported_jlink_runtime
+  RL_CREATE_PLUGIN_MOD_OUTPUT := $(BUILDTOOLS_OUTPUTDIR)/$(RL_BUILD_MODULE_NAME)
+
+  # We need to use FixPath for correct path handling on Windows. Avoid space
+  # in FixPath call in order to avoid needing to use strip.
+  RL_JIMAGE_PATH_ARG := $(call FixPath,$(JDK_LINK_OUTPUT_DIR)/lib/modules)
+  RL_MOD_PATH_ARG := $(call FixPath,$(IMAGES_OUTPUTDIR)/jmods)
+  JLINK_RUNTIME_CREATE_ARG := \
+      -J--module-path=$(RL_CREATE_PLUGIN_MOD_OUTPUT) \
+      -J--add-exports=java.base/jdk.internal.module=$(RL_BUILD_MODULE_NAME) \
+      -J--add-exports=java.base/jdk.internal.jimage=$(RL_BUILD_MODULE_NAME) \
+      -J--add-exports=jdk.jlink/jdk.tools.jlink.internal=$(RL_BUILD_MODULE_NAME) \
+      -J--add-exports=jdk.jlink/jdk.tools.jlink.internal.runtimelink=$(RL_BUILD_MODULE_NAME) \
+      --create-linkable-runtime jimage=$(RL_JIMAGE_PATH_ARG):module-path=$(RL_MOD_PATH_ARG) \
+  #
+
+  # Compile $(RL_BUILD_MODULE_NAME) classes with the 'buildjdk' compiler, as it
+  # needs classes from the just built jdk.jlink module
+  $(eval $(call SetupJavaCompilation, BUILD_JDK_RUNLINK_CLASSES, \
+      COMPILER := buildjdk, \
+      DISABLED_WARNINGS := try, \
```

Comment:
Why do we get warnings in the java code?

Reply:
That's not needed anymore. There are some `try` warnings in the `JmodsReader` and `JimageDiffGenerator` classes.

Reply:
No longer there.